lib/spack/spack/environment.py

```diff
@@ -511,47 +564,78 @@ def add(self, user_spec):
+                speclist.update_reference(new_reference)
         return bool(not existing)
 
-    def remove(self, query_spec, force=False):
+    def remove(self, query_spec, list_name='specs', force=False):
         """Remove specs from an environment that match a query_spec"""
         query_spec = Spec(query_spec)
 
-        # try abstract specs first
-        matches = []
-        if not query_spec.concrete:
-            matches = [s for s in self.user_specs if s.satisfies(query_spec)]
-
-        if not matches:
-            # concrete specs match against concrete specs in the env
-            specs_hashes = zip(
-                self.concretized_user_specs, self.concretized_order)
-            matches = [
-                s for s, h in specs_hashes if query_spec.dag_hash() == h]
-
-        if not matches:
-            raise SpackEnvironmentError("Not found: {0}".format(query_spec))
-
-        for spec in matches:
-            if spec in self.user_specs:
-                self.user_specs.remove(spec)
-
-            if force and spec in self.concretized_user_specs:
-                i = self.concretized_user_specs.index(spec)
-                del self.concretized_user_specs[i]
-
-                dag_hash = self.concretized_order[i]
-                del self.concretized_order[i]
-                del self.specs_by_hash[dag_hash]
+        removed = False
+        for i, (name, speclist) in enumerate(self.read_specs.items()):
+            if name == list_name:
+                # try abstract specs first
+                matches = []
+
+                if not query_spec.concrete:
+                    matches = [s for s in speclist if s.satisfies(query_spec)]
+
+                if not matches:
+                    # concrete specs match against concrete specs in the env
+                    specs_hashes = zip(
+                        self.concretized_user_specs, self.concretized_order)
+                    matches = [
+                        s for s, h in specs_hashes
+                        if query_spec.dag_hash() == h
+                    ]
+
+                if not matches:
+                    raise SpackEnvironmentError(
+                        "Not found: {0}".format(query_spec))
+
+                for spec in matches:
+                    if spec in speclist:
+                        speclist.remove(spec)
+                        removed = True
+
+                    if force and spec in self.concretized_user_specs:
+                        i = self.concretized_user_specs.index(spec)
+                        del self.concretized_user_specs[i]
+
+                        dag_hash = self.concretized_order[i]
+                        del self.concretized_order[i]
+                        del self.specs_by_hash[dag_hash]
+                        removed = True
+
+            elif removed:
+                new_reference = dict((n, self.read_specs[n])
+                                     for n in list(self.read_specs.keys())[:i])
```

Comment:
The reason I update the references for the spec lists inside the loop is that only the spec lists AFTER the one that was modified need to be updated. If I did it outside the loop, it would require a separate iteration over the read_specs

I think your point shows that I need more comments here to clarify what's happening. I'll put comments in both `add` and `remove` to clarify.

Reply:
I think I get your intent but I still think it would be way simpler to do things separately: this `elif`-state-based handling is a headache and in terms of time savings, AFAICT we're saving 50% time on reading ~10 things from a list.

Reply:
@scheibelp It would still have to be handled in an conditional statement of some sort, because I would still have to detect which of the lists need their references updated.

Reply:
It wouldn't be an error if you did this unconditionally would it? I assumed as such because there is a single `removed` switch outside the loop and once it gets flipped each `speclist` is updated. Overall it's making sense to me how this works:

1. you check speclists until you find one with the spec and remove it
1. the elif ensures that you don't modify *that particular list*
1. all future speclists will update their reference (for that matter, if a spec appears in two speclists, is that an error or not and will the reference be updated)

but... IMO it would be simpler to go through removing specs, and then go through removing references: that's a common idiom for handling dangling references in containers.

Reply:
@scheibelp It would not be an error if I did that unconditionally, it would just add a large number of NOPS performance-wise. 

I think I found a way to re-write this that will satisfy your maintainability concerns without having to update references for speclists that don't need it, and is a strict improvement over what I had. Let me know what you think.